### ISSUE TITLE
Encode/decode JS entities works on one byte at a time and is not reversible

### ISSUE THREAD

**Issue by rdipardo** [rdipardo/nppHTMLTag#1]
**[Original report](https://bitbucket.org/rdipardo/htmltag/issues/1) by Anonymous.**

----------------------------------------

- Updated to Notepad 8.3.3 x64
- HtmlTag was missing after update
- Re-installed HtmlTag

Test:
  Following characters require encoding: ä ö ü ß
After encoding:
  Following characters require encoding: &Atilde;&curren; &Atilde;&para; &Atilde;&frac14; &Atilde;&Yuml;
After decoding encoded text:
  Following characters require encoding: Ã¤ Ã¶ Ã¼ ÃŸ

**Comment by rdipardo** [comment]
The decoding algorithm can only handle single-byte sequences. So, this works:

```
\u00E4 \u00F6 \u00FC \u00DF (decode =>) ä ö ü ß
```

But this is broken:

```
ä ö ü ß (encode =>) \u00C3\u00A4 \u00C3\u00B6 \u00C3\u00BC \u00C3\u0178
```

A file in UTF-8 gives 2 bytes to each character, and the algorithm encodes _each one separately_.

That's a limitation of the original author's design \(based on pre-Unicode Notepad\+\+\). It affects both 32- and 64- bit versions.

As I said in  82f9b0e,

> More work still needed before utf8mb4 can be encoded \*correctly\*

Fixing this will be part of that overall task.

**Comment by rdipardo** [comment]
If you're running [at least Windows 10](https://docs.microsoft.com/en-us/answers/questions/587680/where-can-i-find-34beta-use-unicode-utf-8-for-worl.html), here's a way to resolve this issue for the time being:

* Go to the Control Panel, then “Clock and Region”, and select "Change date, time, or number formats"
* Click the "Administrative" tab
* Click "Change system locale..."
* Check the box labelled "Beta: Use Unicode UTF-8 for worldwide language support” \(a reboot is required\)

![](https://bitbucket.org/repo/6749gaq/images/197631860-sys_locale_to_utf8_web.jpg)
‌

Here is N\+\+ 8.3.3 \(64-bit\) on Windows 10 21H2, with the updated system encoding :

‌

![](https://bitbucket.org/repo/6749gaq/images/63746857-htmltagv121.gif)
‌

The plugin is most likely calling a standard library function that uses the system's default encoding. What it _should_ do is encode the document's text as Unicode every time, not rely on Windows.

**Comment by rdipardo** [comment]
Convert selected text to UTF-8 before encoding

Fixes #1

**Comment by rdipardo** [comment]
**Original comment by Björn Klug (Bitbucket: [Björn Klug](https://bitbucket.org/Björn Klug/workspace/repositories)).**

----------------------------------------

I tried your workaround \("Beta: Use Unicode UTF-8 for worldwide language support” checkbox\) but it broke all my MS Access 2010 applications, so that is not a viable solution for me.

Since I’m using this plugin quite frequently I’d by very interested in your estimat when this bug will be fixed.

\[EDIT\] Just found the download of version 1.2.2 at [https://bitbucket.org/rdipardo/htmltag/downloads/](https://bitbucket.org/rdipardo/htmltag/downloads/) which works fine again. Thanks!

**Comment by rdipardo** [comment]
Fixed in d2189a1d3f85c79beceb3237e0484a66a0af0bd0
